https://github.com/haskell/haskell-language-server/commit/097d9902360776079b9218b8b3fd8077b810783a.patch
From: Michael Peyton Jones <me@michaelpj.com>
Date: Sat, 6 Apr 2024 16:01:29 +0100
Subject: [PATCH 1/1] Mark plugins as not buildable if the flag is disabled
 (#4160)

* Mark plugins as not buildable if the flag is disabled

This ensures that cabal does not consider them at all, and won't try to
solve for their dependencies. So if we turn off the fourmolu plugin, cabal
really won't consider fourmolu at all.

This gets us some of the benefits of #4156 with much less work.

Fixes #4100.

* Stick to no space after flag for consistency

Signed-off-by: hololeap <hololeap@protonmail.com>
--- a/haskell-language-server.cabal
+++ b/haskell-language-server.cabal
@@ -116,6 +116,8 @@ flag isolateCabalfmtTests
 
 library hls-cabal-fmt-plugin
   import:           defaults, pedantic, warnings
+  if !flag(cabalfmt)
+    buildable: False
   exposed-modules:  Ide.Plugin.CabalFmt
   hs-source-dirs:   plugins/hls-cabal-fmt-plugin/src
   build-depends:
@@ -132,6 +134,8 @@ library hls-cabal-fmt-plugin
 
 test-suite hls-cabal-fmt-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(cabalfmt)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-cabal-fmt-plugin/test
   main-is:          Main.hs
@@ -161,6 +165,8 @@ common cabal
 
 library hls-cabal-plugin
   import:           defaults, pedantic, warnings
+  if !flag(cabal)
+    buildable: False
   exposed-modules:
     Ide.Plugin.Cabal
     Ide.Plugin.Cabal.Diagnostics
@@ -204,6 +210,8 @@ library hls-cabal-plugin
 
 test-suite hls-cabal-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(cabal)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-cabal-plugin/test
   main-is:          Main.hs
@@ -242,6 +250,8 @@ common class
 
 library hls-class-plugin
   import:           defaults, pedantic, warnings
+  if !flag(class)
+    buildable: False
   exposed-modules:    Ide.Plugin.Class
   other-modules:      Ide.Plugin.Class.CodeAction
                     , Ide.Plugin.Class.CodeLens
@@ -272,6 +282,8 @@ library hls-class-plugin
 
 test-suite hls-class-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(class)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-class-plugin/test
   main-is:          Main.hs
@@ -301,7 +313,8 @@ common callHierarchy
 
 library hls-call-hierarchy-plugin
   import: defaults, pedantic, warnings
-  buildable: True
+  if !flag(callHierarchy)
+    buildable: False
   exposed-modules:    Ide.Plugin.CallHierarchy
   other-modules:
     Ide.Plugin.CallHierarchy.Internal
@@ -326,6 +339,8 @@ library hls-call-hierarchy-plugin
 
 test-suite hls-call-hierarchy-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(callHierarchy)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-call-hierarchy-plugin/test
   main-is:          Main.hs
@@ -359,6 +374,8 @@ common eval
 
 library hls-eval-plugin
   import: defaults, pedantic, warnings
+  if !flag(eval)
+    buildable: False
   exposed-modules:
     Ide.Plugin.Eval
     Ide.Plugin.Eval.Types
@@ -405,6 +422,8 @@ library hls-eval-plugin
 
 test-suite hls-eval-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(eval)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-eval-plugin/test
   main-is:          Main.hs
@@ -439,6 +458,8 @@ flag importLens
 
 library hls-explicit-imports-plugin
   import:           defaults, pedantic, warnings
+  if !flag(importLens)
+    buildable: False
   exposed-modules:  Ide.Plugin.ExplicitImports
   hs-source-dirs:   plugins/hls-explicit-imports-plugin/src
   build-depends:
@@ -461,6 +482,8 @@ library hls-explicit-imports-plugin
 
 test-suite hls-explicit-imports-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(importLens)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-explicit-imports-plugin/test
   main-is:          Main.hs
@@ -491,6 +514,8 @@ common rename
 
 library hls-rename-plugin
   import:           defaults, pedantic, warnings
+  if !flag(rename)
+    buildable: False
   exposed-modules:  Ide.Plugin.Rename
   hs-source-dirs:   plugins/hls-rename-plugin/src
   build-depends:
@@ -515,6 +540,8 @@ library hls-rename-plugin
 
 test-suite hls-rename-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(rename)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-rename-plugin/test
   main-is:          Main.hs
@@ -543,6 +570,8 @@ common retrie
 
 library hls-retrie-plugin
   import:           defaults, pedantic, warnings
+  if !flag(retrie)
+    buildable: False
   exposed-modules:  Ide.Plugin.Retrie
   hs-source-dirs:   plugins/hls-retrie-plugin/src
   build-depends:
@@ -573,6 +602,8 @@ library hls-retrie-plugin
 
 test-suite hls-retrie-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(retrie)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-retrie-plugin/test
   main-is:          Main.hs
@@ -601,6 +632,8 @@ common hlint
 
 library hls-hlint-plugin
   import:           defaults, pedantic, warnings
+  if !flag(hlint)
+    buildable: False
   exposed-modules:  Ide.Plugin.Hlint
   hs-source-dirs:   plugins/hls-hlint-plugin/src
   build-depends:
@@ -635,6 +668,8 @@ library hls-hlint-plugin
 
 test-suite hls-hlint-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(hlint)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-hlint-plugin/test
   main-is:          Main.hs
@@ -667,7 +702,7 @@ common stan
 
 library hls-stan-plugin
   import:           defaults, pedantic, warnings
-  if (impl(ghc > 8.8.1) && impl(ghc <= 9.2.3) || impl(ghc >= 9.4.0) && impl(ghc < 9.10.0))
+  if flag(stan) && (impl(ghc > 8.8.1) && impl(ghc <= 9.2.3) || impl(ghc >= 9.4.0) && impl(ghc < 9.10.0))
     buildable: True
   else
     buildable: False
@@ -695,7 +730,7 @@ library hls-stan-plugin
 
 test-suite hls-stan-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
-  if (impl(ghc > 8.8.1) && impl(ghc <= 9.2.3) || impl(ghc >= 9.4.0) && impl(ghc < 9.10.0))
+  if flag(stan) && (impl(ghc > 8.8.1) && impl(ghc <= 9.2.3) || impl(ghc >= 9.4.0) && impl(ghc < 9.10.0))
     buildable: True
   else
     buildable: False
@@ -730,6 +765,8 @@ common moduleName
 
 library hls-module-name-plugin
   import:           defaults, pedantic, warnings
+  if !flag(moduleName)
+    buildable: False
   exposed-modules:  Ide.Plugin.ModuleName
   hs-source-dirs:   plugins/hls-module-name-plugin/src
   build-depends:
@@ -747,6 +784,8 @@ library hls-module-name-plugin
 
 test-suite hls-module-name-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(moduleName)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-module-name-plugin/test
   main-is:          Main.hs
@@ -772,6 +811,8 @@ common pragmas
 
 library hls-pragmas-plugin
   import:           defaults, pedantic, warnings
+  if !flag(pragmas)
+    buildable: False
   exposed-modules:  Ide.Plugin.Pragmas
   hs-source-dirs:   plugins/hls-pragmas-plugin/src
   build-depends:
@@ -788,6 +829,8 @@ library hls-pragmas-plugin
 
 test-suite hls-pragmas-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(pragmas)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-pragmas-plugin/test
   main-is:          Main.hs
@@ -817,6 +860,8 @@ common splice
 
 library hls-splice-plugin
   import:           defaults, pedantic, warnings
+  if !flag(splice)
+    buildable: False
   exposed-modules:
     Ide.Plugin.Splice
     Ide.Plugin.Splice.Types
@@ -845,6 +890,8 @@ library hls-splice-plugin
 
 test-suite hls-splice-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(splice)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-splice-plugin/test
   main-is:          Main.hs
@@ -872,6 +919,8 @@ common alternateNumberFormat
 
 library hls-alternate-number-format-plugin
   import:           defaults, pedantic, warnings
+  if !flag(alternateNumberFormat)
+    buildable: False
   exposed-modules:  Ide.Plugin.AlternateNumberFormat, Ide.Plugin.Conversion
   other-modules:    Ide.Plugin.Literals
   hs-source-dirs:   plugins/hls-alternate-number-format-plugin/src
@@ -897,6 +946,8 @@ library hls-alternate-number-format-plugin
 
 test-suite hls-alternate-number-format-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(alternateNumberFormat)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-alternate-number-format-plugin/test
   other-modules:    Properties.Conversion
@@ -932,6 +983,8 @@ common qualifyImportedNames
 
 library hls-qualify-imported-names-plugin
   import:           defaults, pedantic, warnings
+  if !flag(qualifyImportedNames)
+    buildable: False
   exposed-modules:  Ide.Plugin.QualifyImportedNames
   hs-source-dirs:   plugins/hls-qualify-imported-names-plugin/src
   build-depends:
@@ -950,6 +1003,8 @@ library hls-qualify-imported-names-plugin
 
 test-suite hls-qualify-imported-names-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(qualifyImportedNames)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-qualify-imported-names-plugin/test
   main-is:          Main.hs
@@ -976,6 +1031,8 @@ common codeRange
 
 library hls-code-range-plugin
   import:           defaults, pedantic, warnings
+  if !flag(codeRange)
+    buildable: False
   exposed-modules:
     Ide.Plugin.CodeRange
     Ide.Plugin.CodeRange.Rules
@@ -999,6 +1056,8 @@ library hls-code-range-plugin
 
 test-suite hls-code-range-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(codeRange)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-code-range-plugin/test
   main-is:          Main.hs
@@ -1033,6 +1092,8 @@ common changeTypeSignature
 
 library hls-change-type-signature-plugin
   import:           defaults, pedantic, warnings
+  if !flag(changeTypeSignature)
+    buildable: False
   exposed-modules:  Ide.Plugin.ChangeTypeSignature
   hs-source-dirs:   plugins/hls-change-type-signature-plugin/src
   build-depends:
@@ -1054,6 +1115,8 @@ library hls-change-type-signature-plugin
 
 test-suite hls-change-type-signature-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(changeTypeSignature)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-change-type-signature-plugin/test
   main-is:          Main.hs
@@ -1084,6 +1147,8 @@ common gadt
 
 library hls-gadt-plugin
   import:           defaults, pedantic, warnings
+  if !flag(gadt)
+    buildable: False
   exposed-modules:  Ide.Plugin.GADT
   other-modules:    Ide.Plugin.GHC
   hs-source-dirs:   plugins/hls-gadt-plugin/src
@@ -1107,6 +1172,8 @@ library hls-gadt-plugin
 
 test-suite hls-gadt-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(gadt)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-gadt-plugin/test
   main-is:          Main.hs
@@ -1133,6 +1200,8 @@ common explicitFixity
 
 library hls-explicit-fixity-plugin
   import:           defaults, pedantic, warnings
+  if !flag(explicitFixity)
+    buildable: False
   exposed-modules:  Ide.Plugin.ExplicitFixity
   hs-source-dirs:   plugins/hls-explicit-fixity-plugin/src
   build-depends:
@@ -1150,6 +1219,8 @@ library hls-explicit-fixity-plugin
 
 test-suite hls-explicit-fixity-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(explicitFixity)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-explicit-fixity-plugin/test
   main-is:          Main.hs
@@ -1176,6 +1247,8 @@ common explicitFields
 
 library hls-explicit-record-fields-plugin
   import:           defaults, pedantic, warnings
+  if !flag(explicitFields)
+    buildable: False
   exposed-modules:  Ide.Plugin.ExplicitFields
   build-depends:
     , base                  >=4.12 && <5
@@ -1196,6 +1269,8 @@ library hls-explicit-record-fields-plugin
 
 test-suite hls-explicit-record-fields-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(explicitFields)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-explicit-record-fields-plugin/test
   main-is:          Main.hs
@@ -1222,6 +1297,8 @@ common overloadedRecordDot
 
 library hls-overloaded-record-dot-plugin
   import:           defaults, pedantic, warnings
+  if !flag(overloadedRecordDot)
+    buildable: False
   exposed-modules:  Ide.Plugin.OverloadedRecordDot
   build-depends:
     , base                  >=4.16 && <5
@@ -1240,6 +1317,8 @@ library hls-overloaded-record-dot-plugin
 
 test-suite hls-overloaded-record-dot-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(overloadedRecordDot)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-overloaded-record-dot-plugin/test
   main-is:          Main.hs
@@ -1267,6 +1346,8 @@ common floskell
 
 library hls-floskell-plugin
   import:           defaults, pedantic, warnings
+  if !flag(floskell)
+    buildable: False
   exposed-modules:  Ide.Plugin.Floskell
   hs-source-dirs:   plugins/hls-floskell-plugin/src
   build-depends:
@@ -1281,6 +1362,8 @@ library hls-floskell-plugin
 
 test-suite hls-floskell-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(floskell)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-floskell-plugin/test
   main-is:          Main.hs
@@ -1306,6 +1389,8 @@ common fourmolu
 
 library hls-fourmolu-plugin
   import:           defaults, pedantic, warnings
+  if !flag(fourmolu)
+    buildable: False
   exposed-modules:  Ide.Plugin.Fourmolu
   hs-source-dirs:   plugins/hls-fourmolu-plugin/src
   build-depends:
@@ -1325,6 +1410,8 @@ library hls-fourmolu-plugin
 
 test-suite hls-fourmolu-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(fourmolu)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-fourmolu-plugin/test
   main-is:          Main.hs
@@ -1355,6 +1442,8 @@ common ormolu
 
 library hls-ormolu-plugin
   import:           defaults, pedantic, warnings
+  if !flag(ormolu)
+    buildable: False
   exposed-modules:  Ide.Plugin.Ormolu
   hs-source-dirs:   plugins/hls-ormolu-plugin/src
   build-depends:
@@ -1374,6 +1463,8 @@ library hls-ormolu-plugin
 
 test-suite hls-ormolu-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(ormolu)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-ormolu-plugin/test
   main-is:          Main.hs
@@ -1405,6 +1496,8 @@ common stylishHaskell
 
 library hls-stylish-haskell-plugin
   import:           defaults, pedantic, warnings
+  if !flag(stylishHaskell)
+    buildable: False
   exposed-modules:  Ide.Plugin.StylishHaskell
   hs-source-dirs:   plugins/hls-stylish-haskell-plugin/src
   build-depends:
@@ -1422,6 +1515,8 @@ library hls-stylish-haskell-plugin
 
 test-suite hls-stylish-haskell-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(stylishHaskell)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-stylish-haskell-plugin/test
   main-is:          Main.hs
@@ -1447,6 +1542,8 @@ common refactor
 
 library hls-refactor-plugin
   import:           defaults, pedantic, warnings
+  if !flag(refactor)
+    buildable: False
   exposed-modules:  Development.IDE.GHC.ExactPrint
                     Development.IDE.GHC.Compat.ExactPrint
                     Development.IDE.Plugin.CodeAction
@@ -1504,6 +1601,8 @@ library hls-refactor-plugin
 
 test-suite hls-refactor-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(refactor)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-refactor-plugin/test
   main-is:          Main.hs
@@ -1547,7 +1646,8 @@ common semanticTokens
 
 library hls-semantic-tokens-plugin
   import: defaults, pedantic, warnings
-  buildable: True
+  if !flag(semanticTokens)
+    buildable: False
   exposed-modules:
     Ide.Plugin.SemanticTokens
     Ide.Plugin.SemanticTokens.Types
@@ -1587,6 +1687,8 @@ library hls-semantic-tokens-plugin
 
 test-suite hls-semantic-tokens-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
+  if !flag(semanticTokens)
+    buildable: False
   type:             exitcode-stdio-1.0
   hs-source-dirs:   plugins/hls-semantic-tokens-plugin/test
   main-is:          SemanticTokensTest.hs
-- 
2.43.2

